Add support for Taproot and `tr()` descriptors

afilini opened the pull request:
### Description

This is a work-in-progress PR to update BDK to rust-bitcoin `0.28` which introduces taproot support and a few other improvements. While updating we also introduce taproot support in BDK.

High level list of subtasks for this PR:
- [x] Update rust-bitcoin and rust-miniscript
- [x] Stop using deprecated structs
- [x] Add taproot metadata to psbts
- [x] Produce schnorr signatures
- [x] Finalize taproot txs
- [x] Support `tr()` descriptors in the `descriptor!()` macro
- [x] Write a lot of tests
  - [x] Interoperability with other wallets (Core + ?)
     - [x] Signing/finalizing a psbt made by core
     - [x] Producing a psbt that core can sign and finalize
  - [x] Creating psbts
    - [x] Verify the metadata are correct
    - [x] Verify sighashes are applied correctly
    - [x] Create a tx with a foreign taproot and non-taproot utxo
  - [x] Signing psbts
    - [x] Signing for a key spend
    - [x] Signing for a script spend
    - [x] Signing with a single (wif) key
    - [x] Signing with an xprv (with and without knowing the utxo being spent in the db)
    - [x] Signing with weird sighashes
  - [x] Policy module
    - [x] Simple key spend
    - [x] More complex tap tree with a few keys
    - [x] Verify both `contribution` and `satisfaction` of a PSBT input
  - [x] Wallet module
    - [x] Generate addresses

Fixes #63 

### Notes to the reviewers

#### Milestone

I'm adding this to the `0.19` milestone because now that rust-bitcoin and rust-miniscript have been released we should not waiting too long to release a version of BDK that supports the new libraries.

#### API Breaks

Since this is an API-break because of the new version of rust-bitcoin and rust-miniscript, I'm also taking the chance to update a few things in our lib that I had been thinking about for a while.

One example is the signer interface, which had that weird `sign_whole_tx()` method. This has now been removed, and the `Signer` trait replaced with `TransactionSigner` and `InputSigner`. I'm also starting to think that the signer should not only look at the psbt to figure out what to do, but ideally it should also receive some information about the descriptor (for example, the type) to simplify the code.

One option is to add an extra parameter, but that would probably only be used by our internal signers and not much else (for example, if you ask an hardware wallet to sign, it will probably already know what kind of wallet you have).

Another option is to wrap `PrivateKey` and `DescriptorXKey<ExtendedPrivKey>` which are the two internal signers we support with a struct that contains metadata about the descriptor, and then implement the signer traits on that struct. We could construct this in `Wallet::new()`, after miniscript parses the descriptor.

#### MSRV Bump

Due to the update of `rust-electrum-client`, which in turn depends on an updated `webpki`, we will have to bump our MSRV beacuse 1.46 is not supported by the new `webpki` version.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

afilini committed:
Add `tr()` descriptors to the `descriptor!()` macro

afilini committed:
Refactor signer traits, add support for taproot signatures

afilini committed:
Populate `tap_key_origin` in PSBT inputs and outputs

afilini committed:
Populate more taproot fields in PSBTs

afilini committed:
Use `tap_key_origins` in PSBTs to derive descriptors

afilini committed:
Support `tr()` descriptors in dsl

afilini committed:
Fix type inference for the `tr()` descriptor, add basic tests

afilini commented:
I've updated the PR description with a list of ideas for tests that we need for this. The list is non-exhaustive, just a point from which we can start.

notmandatory commented:
I've started looking into the "Creating psbts" tests. 

Some notes on setting up a local `bitcoind -regtest` taproot wallet here: https://hackmd.io/@notmandatory/Sy2yOYeI5

rajarshimaitra commented:
I am finding a lot of version conflict while working on this https://github.com/bitcoindevkit/bdk-cli/issues/62#issuecomment-1124655758 while trying to update bdk-cli with the latest `bitcoincore-rpc` and `electrsd`.. 

I think the first two commit of this PR if can be merged independently of taproot into master would be really helpful.. 

It seems recently a lot has changed in the upstreams and we are using `bitcoincore-rpc` and `electrum-client` from two different sources (directly, and then from `electrsd`). And when we try to move `electrsd` from dev dependency all sort of version conflict is appearing in bdk-cli..

But to sort them in bdk-cli side I need to have the final updated version of Cargo.lock in bdk, which is done by the first two commits..

rajarshimaitra commented:
It also might make sense if we just make a patch release with the dependency upgrade, then the downstream libraries can fix them up. Or just having them in master would work too for time being.

afilini commented:
I don't think it can be released as patch upgrade because it would be a breaking change. I.e. if you depend on bdk `0.18` you would get the update and it would break your code.

I think I agree, I'm gonna open a new PR now with just the upgrade

afilini committed:
Attach a context to our software signers

This allows the signer to know the signing context precisely without
relying on heuristics on the psbt fields.

Due to the context being static, we still have to look at the PSBT when
producing taproot signatures to determine the set of leaf hashes that
the key can sign for.

afilini commented:
As discussed in our last meeting on Discord, I've added commit ea165b02bf7698242b8c7ff5485b23331c2ac832 to introduce the concept of a `SignerContext`, which helps the signer produce the right signature without having to rely on specific fields of the PSBT being set.

Unfortunately since the context is static (reused for all the signatures of the same signer) I couldn't store the list of taproot leaf hashes in there, because those change with every derivation index. So I only store whether the key matches the "internal_key", but then we still rely on the PSBT to figure out the list of leaf hashes to sign for.

notmandatory commented:
I'm working on some commits to add to this PR with tests for "Interoperability with other wallets ", see: https://github.com/afilini/bdk/pull/4. So far only testing with core (22.0).

afilini commented:
I'm working on the "signing psbts" section

afilini committed:
taproot-tests: Test taproot key and script spend in the blockchain tests

This is to ensure a Bitcoin node accepts our transactions

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -301,48 +384,122 @@ impl Signer for PrivateKey {
+) {
+    use schnorr::TapTweak;
+
+    let keypair = secp256k1::KeyPair::from_seckey_slice(secp, secret_key.as_ref()).unwrap();
+    let keypair = match leaf_hash {
+        None => keypair
+            .tap_tweak(secp, psbt_input.tap_merkle_root)
+            .into_inner(),
+        Some(_) => keypair, // no tweak for script spend
+    };
```
What if the secret key given here doesn't match with the pubkey?? Better to make a proactive check for that as we have the pubkey as an input already??

afilini replied:
I would skip the check here considering that we already have some of them in the `InputSigner` implementation, and an extra ec multiplication can be a bit slower.

Worst case we would just produce an invalid signature, or place a valid signature in the wrong spot.

rajarshimaitra replied:
Okay ya makes sense, as long as we have some checks elsewhere..

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -301,48 +384,122 @@ impl Signer for PrivateKey {
+fn produce_schnorr_sig(
+    secret_key: &secp256k1::SecretKey,
+    pubkey: XOnlyPublicKey,
+    leaf_hash: Option<taproot::TapLeafHash>,
```
A comment somewhere here saying making the `leaf_hash` `None` will use the signature for key path spend, would be helpful. Even if just for internal use, because I will surely forget this.. 😅

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -229,19 +299,25 @@ impl Signer for DescriptorXKey<ExtendedPrivKey> {
-            Some((pk, full_path)) => (pk, full_path.clone()),
+            Some((pk, full_path)) => (pk, full_path),
             None => return Ok(()),
         };
```
Can't we just `collect()` them out here instead of `next() { ... }` block?? Found this hard to parse through..

afilini replied:
Collecting allocates a new vec but we really only need the first item, so that would be a bit more wasteful.

I'll see if I can write this down more clearly.

afilini replied:
Ok I switched to `find_map()` which directly returns the first non-`None` item, instead of mapping the iterator and taking the first item.

rajarshimaitra replied:
Yes.. Thats even more better..

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -301,48 +384,122 @@ impl Signer for PrivateKey {
-    fn sign_whole_tx(&self) -> bool {
-        false
-    }
+fn produce_ecsda_sig(
```
Nit: because these functions modifies a psbt and not produce any signature, maybe they are better named as `sign_psbt_schnorr` or something?? 

Nothing blocking, may consider next time you touch this..

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -631,7 +822,62 @@ impl ComputeSighash for Segwitv0 {
+            return Err(SignerError::MissingWitnessUtxo);
+        };
+
+        // Assume no OP_CODESEPARATOR
+        let extra = extra.map(|leaf_hash| (leaf_hash, 0xFFFFFFFF));
```
What happens when there are code separators?? I am still cloudy on where does the code separator comes from and what it does.. As far as I can figure out its somewhere in the witness script itself?? So we have to count them out and then input it here??

afilini replied:
If there are codeseparators and we don't account for them we may produce an invalid signature.

However, I haven't seen a single example script for taproot using op_codeseparator. The BIP mentions that there may be one very specific use case, but it doesn't even show how to build it.

Miniscript for sure doesn't use them, so as long as we only support miniscript descriptors we should be fine. We may have to reconsider this after #485, because then a user could make a custom script with the codeseparator, but I wouldn't worry too much about it.

rajarshimaitra replied:
Ya that makes sense.. If its not that much useful no point worrying much on it.. I also haven't seen any codeseperator in the wild.. Maybe we can make a note somewhere saying BDK doesn't support it yet..

rajarshimaitra reviewed (commented):
Review ACK 8b709a595dc0da324b8cd357fbf4345f280b8bb9

Below are some non blocking comments and questions.. 

It seems the `wallet::signer` module can use some unit tests?? Is that necessary or they are covered elsewhere??

afilini committed:
taproot-tests: Add test coverage for tx signing

rajarshimaitra commented on src/wallet/signer.rs:
```diff
@@ -174,27 +174,46 @@ impl fmt::Display for SignerError {
+    /// Taproot context (BIP 340)
+    Tap {
+        /// Whether the signer can sign for the internal key or not
+        internal_key: bool,
```
Nit: Can it be named `is_internal_key` (intends a flag) instead??

rajarshimaitra reviewed (commented):
ACK 9034096089cb1067ad1a3bde94f43f74ed8dd2b2

just one nit..

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -1494,17 +1503,21 @@ where
 
```
Unrelated to taproot, but in the lines above, we are deriving the `KeyChainKind` and `Child` index for the spk.. The `KeyChainKind` is also present in `LocalUtxo`. So is it better to make a match that these two are same? Or its always guaranted to be the same?

afilini replied:
In general that kind of mismatch should only happen if the database is inconsistent, or if the user is actively manipulating the data.

Besides, all of the code right now assumes the database is the ultimate source of truth, so I think it's fine here to ignore the value of `LocalUtxo` and just use what comes out of it.

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -1496,17 +1528,35 @@ where
+            psbt_input.bip32_derivation = derived_descriptor.get_hd_keypaths(&self.secp);
+        }
 
         psbt_input.redeem_script = derived_descriptor.psbt_redeem_script();
         psbt_input.witness_script = derived_descriptor.psbt_witness_script();
```
It seems to me that in these two cases we are not considering `Wpkh` types?? What happens for them??

https://github.com/bitcoindevkit/bdk/blob/4c1749a13a29da6be9812948f6aefb1ed74f166f/src/descriptor/mod.rs#L330-L349

afilini replied:
The `redeem_script` field is only populated for P2SH and P2WSH address types, so here we ignore all the others (P2TR, P2PKH, P2WPKH)

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -4097,4 +4170,167 @@ pub(crate) mod test {
+            psbt.outputs[0].tap_key_origins.len(),
+            1,
+            "Empty output tap_key_origins"
+        );
```
Instead of only checking length feels like its better to assert the data structure itself?? The descriptor is hardcoded so they should not change..

afilini replied:
Yes, that's a good point

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -4097,4 +4170,167 @@ pub(crate) mod test {
+            "Empty input tap_merkle_root"
+        );
+        assert_eq!(psbt.inputs[0].tap_scripts.len(), 2, "Empty tap_scripts");
+    }
```
Same test comment on asserting the data than checking length..

rajarshimaitra reviewed (commented):
tACK 58d5abc5225572fedd641644456f09fca31879d3

Few nits and comments..

afilini commented:
I'm now working on the policy module tests, I will probably include some for the descriptor macro as well which were not part of the original list

afilini committed:
policy: Refactor `PkOrF` into an enum

For whatever reason we were using a struct as an enum, so we might as
well fix it in this PR since we are already breaking the API quite
badly.

afilini committed:
policy: Build `SatisfiableItem::*Signature` based on the context

Also refactor our code to lookup signatures in PSBTs to use the context

afilini committed:
taproot-tests: Add tests for the policy module

afilini committed:
policy: Consider `tap_key_origins` when looking for sigs in PSBTs

We used to only look at `bip32_derivations` which is only used for ECDSA
keys.

notmandatory committed:
Add blockchain tests for parsing, signing, finalizing taproot core psbts

notmandatory commented:
When you get a chance this needs a rebase to pick up 35feb107ed5969720ce54a6aa76b7b2176f6c7c1 to get CI to run the blockchain tests correctly.

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -1436,8 +1437,10 @@ where
                 } => {
-                    // TODO: do not require non_witness_utxo for taproot utxos
-                    if !params.only_witness_utxo && foreign_psbt_input.non_witness_utxo.is_none() {
+                    if foreign_psbt_input.tap_internal_key.is_none()
```
Can you add a small comment explaining why it's ok not having the `non_witness_utxo` if you don't have the tap internal key? Even a small comment with a URL to the doc is ok

danielabrozzoni commented on src/wallet/signer.rs:
```diff
@@ -631,7 +760,62 @@ impl ComputeSighash for Segwitv0 {
+
+        let mut cache = sighash::SighashCache::new(&psbt.unsigned_tx);
+        let is_anyone_can_pay = psbt::PsbtSighashType::from(sighash_type).to_u32() & 0x80 != 0;
+        let prevouts = if is_anyone_can_pay {
+            sighash::Prevouts::One(
+                input_index,
+                witness_utxos[input_index]
+                    .as_ref()
+                    .ok_or(SignerError::MissingWitnessUtxo)?,
+            )
+        } else if witness_utxos.iter().all(Option::is_some) {
+            all_witness_utxos.extend(witness_utxos.iter().filter_map(|x| x.as_ref()));
+            sighash::Prevouts::All(&all_witness_utxos)
+        } else {
```
TODO: test with ANYONECANPAY, ALL and with missing witness utxos

danielabrozzoni commented on src/descriptor/mod.rs:
```diff
@@ -515,7 +523,64 @@ impl<'s> DerivedDescriptorMeta for DerivedDescriptor<'s> {
+
+            // If this is the internal key, we only insert the key origin if it's not None.
+            // For keys found in the tap tree we always insert a key origin (because the signer
+            // looks for it to know which leaves to sign for), but the key origin may be empty
```
nit: I would rewrite this as
```
 // For keys found in the tap tree we always insert a key origin (because the signer
 // looks for it to know which leaves to sign for), even though it may be None
```

It's probably only me, but when I read it the first time I thought there was some kinda typo as you were repeating "key origin" twice 😅  Feel free to ignore.

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -1461,7 +1462,15 @@ where
+                        .tap_key_origins
+                        .append(&mut derived_descriptor.get_tap_key_origins(&self.secp));
+                } else {
+                    psbt_output
+                        .bip32_derivation
+                        .append(&mut derived_descriptor.get_hd_keypaths(&self.secp));
```
You were assigning `derived_descriptor.get_hd_keypaths(&self.secp)` to `psbt_output.bip32_derivation` before, now you're appending. Is there a specific reason for doing so? I'm trying to understand if I'm missing something 😅

afilini replied:
Appending doesn't override what's already there. So let's say a wallet creates the PSBT with some key origins that we don't know of, then while we sign we would delete them, which would potentially make it impossible for a third wallet to sign if it needed those specific origins.

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -1013,23 +1014,27 @@ where
-        // If we aren't allowed to use `witness_utxo`, ensure that every input but finalized one
-        // has the `non_witness_utxo`
+        // If we aren't allowed to use `witness_utxo`, ensure that every input (except p2tr and finalized ones)
+        // have the `non_witness_utxo`
```
nit: has?

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -1013,23 +1014,27 @@ where
+                i.sighash_type.is_none()
+                    || i.sighash_type == Some(EcdsaSighashType::All.into())
+                    || i.sighash_type == Some(SchnorrSighashType::All.into())
+                    || i.sighash_type == Some(SchnorrSighashType::Default.into())
```
Does this mean we don't return an error here if we have an input spending non-taproot outputs with `SchnorrSighashType::Default`?

afilini replied:
That's true, although it will fail later on in the signer code when it tires to "cast" the `PsbtSighashType` to an `EcdsaSighashType`.

Checking that you are using a "compatible" sighash type in my opinion falls in the "integrity check" kind of validation, which we don't really do right now in BDK. Maybe we should start, but that's not in the scope of this PR I think

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -4095,4 +4130,39 @@ pub(crate) mod test {
     }
+
+    #[test]
+    fn test_taproot_psbt_populate_tap_key_origins() {
```
In this test we are testing the script path key origin insertion.. Probably that's because you didn't had the script path tr descriptor at that time.. But now we can test script path key origin insertion using `get_test_tr_with_taptree_xprv` it seems?

afilini replied:
It's kinda testing two different things, but I agree it makes sense to test that part as well. Here we are testing the key origins for key spend (because the leaf hash vec is empty), I'll add one more test for when the key appears in multiple leaves to make sure we are accounting for all of them

rajarshimaitra reviewed (commented):
tACK f2f2d56ab77578ce1a6746ed30267cd3a3628676

Just one test suggestion..

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -4165,4 +4203,35 @@ pub(crate) mod test {
+            vec![
+                (taproot::ControlBlock::from_slice(&Vec::<u8>::from_hex("c151494dc22e24a32fe9dcfbd7e85faf345fa1df296fb49d156e859ef3452012958b0afded0ee3149dbf6710d349dc30e55ae30c319c30efbc79efe19cf70f46a8").unwrap()).unwrap(), (from_str!("208aee2b8120a5f157f1223f72b5e62b825831a27a9fdf427db7cc697494d4a642ac"), taproot::LeafVersion::TapScript)),
+                (taproot::ControlBlock::from_slice(&Vec::<u8>::from_hex("c151494dc22e24a32fe9dcfbd7e85faf345fa1df296fb49d156e859ef345201295b9a515f7be31a70186e3c5937ee4a70cc4b4e1efe876c1d38e408222ffc64834").unwrap()).unwrap(), (from_str!("20b511bd5771e47ee27558b1765e87b541668304ec567721c7b880edc0a010da55ac"), taproot::LeafVersion::TapScript)),
+            ],
```
Might make sense to also assert the `psbt_input.tap_internal_key` and `psbt_input.tap_merkle_root`?? 
Also same for the `psbt.output` taproot data..

afilini replied:
The merkle root is asserted a few lines above it. I'll change the test to send back to the same address as the input so that I can then assert that the input metadata are equal to the output

afilini replied:
I just realized that these metadata are not present in the output, so there's not much to check there

rajarshimaitra reviewed (commented):
tACK 14d86623ca75c4be3f633d2495d96d2ac3f52535

danielabrozzoni reviewed (commented):
Concept ACK :)
I have a couple of comments, mainly regarding making the code easier to understand.
I still haven't looked into tests

rajarshimaitra commented on src/descriptor/mod.rs:
```diff
@@ -393,61 +411,124 @@ impl DescriptorMeta for ExtendedDescriptor {
             if let DescriptorPublicKey::XPub(xpub) = key.as_key().deref() {
-                // Check if the key matches one entry in our `index`. If it does, `matches()` will
+                // Check if the key matches one entry in our `key_origins`. If it does, `matches()` will
                 // return the "prefix" that matched, so we remove that prefix from the full path
-                // found in `index` and save it in `derive_path`. We expect this to be a derivation
+                // found in `key_origins` and save it in `derive_path`. We expect this to be a derivation
                 // path of length 1 if the key is `wildcard` and an empty path otherwise.
```
What happens when the `derive_path` is more than one childnumber long?? Should we assert it somewhere and throw an error as unexpected behavior?? There is no restriction on the input derivation path in key origin to be only extending it by one childnumber.. 

Right now I think if the input diverges from the expectation the `if let` statement at the end will be skipped and it won't return anything.. But that can be confusing to users..

afilini replied:
This is actually guaranteed by miniscript, the `matches()` function returns a path of either length one or zero (see https://docs.rs/miniscript/7.0.0/src/miniscript/descriptor/key.rs.html#672).

rajarshimaitra reviewed (commented):
tACK 6fa5832a321b3596d726277107df59685e2904d9

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -4097,4 +4185,269 @@ pub(crate) mod test {
+
+    #[test]
+    fn test_taproot_sign_derive_index_from_psbt() {
+        env_logger::init();
```
Debugging leftover?

danielabrozzoni commented on src/wallet/mod.rs:
```diff
@@ -4097,4 +4185,269 @@ pub(crate) mod test {
+    }
+
+    #[test]
+    fn test_taproot_psbt_input_tap_tree() {
```
Added a check for the tap tree:
```
diff --git a/src/wallet/mod.rs b/src/wallet/mod.rs
index 344bf11..8f7510f 100644
--- a/src/wallet/mod.rs
+++ b/src/wallet/mod.rs
@@ -4284,6 +4284,8 @@ pub(crate) mod test {
 
     #[test]
     fn test_taproot_psbt_input_tap_tree() {
+        use crate::bitcoin::psbt::serialize::Deserialize;
+        use crate::bitcoin::psbt::TapTree;
         use bitcoin::hashes::hex::FromHex;
         use bitcoin::util::taproot;
 
@@ -4323,6 +4325,11 @@ pub(crate) mod test {
             psbt.inputs[0].tap_internal_key,
             psbt.outputs[0].tap_internal_key
         );
+
+        assert_eq!(
+            psbt.outputs[0].tap_tree,
+            Some(TapTree::deserialize(&Vec::<u8>::from_hex("01c022208aee2b8120a5f157f1223f72b5e62b825831a27a9fdf427db7cc697494d4a642ac01c0222051494dc22e24a32fe9dcfbd7e85faf345fa1df296fb49d156e859ef345201295ac",).unwrap()).unwrap())
+        );
     }
 
     fn test_spend_from_wallet(wallet: Wallet<AnyDatabase>) {
```

rajarshimaitra commented on src/descriptor/dsl.rs:
```diff
@@ -640,21 +764,22 @@ macro_rules! fragment {
+        $crate::keys::make_multi($thresh, $crate::miniscript::Terminal::Multi, $keys, &secp)
     });
     ( multi ( $thresh:expr $(, $key:expr )+ ) ) => ({
-        use $crate::keys::IntoDescriptorKey;
+        $crate::group_multi_keys!( $( $key ),* )
```
Probably I am missing something, but why the input `key` pattern is `1 or many` and the pattern in code is `0 or many`? (the `+` in input and `*` in the code after `key`)

afilini replied:
I think it doesn't matter what you use when expanding repeated items, the compiler only looks at what you have in the "declaration" to match the tokens passed to the macro.

So in this case it should error if there's no key at all (because of the `+`), but afterwards you just want to iterate over whatever you have and for that you can just use `*`.

rajarshimaitra replied:
Ok, clear now.. Didn't knew that.. Thanks..

rajarshimaitra commented on src/descriptor/dsl.rs:
```diff
@@ -640,21 +764,22 @@ macro_rules! fragment {
-            .and_then(|keys| $crate::keys::make_multi($thresh, keys, &secp))
+        $crate::keys::make_multi($thresh, $crate::miniscript::Terminal::MultiA, $keys, &secp)
+    });
+    ( multi_a ( $thresh:expr $(, $key:expr )+ ) ) => ({
+        $crate::group_multi_keys!( $( $key ),* )
```
Similar question on key pattern..

rajarshimaitra commented on src/keys/mod.rs:
```diff
@@ -792,13 +792,18 @@ pub fn make_pkh<Pk: IntoDescriptorKey<Ctx>, Ctx: ScriptContext>(
+pub fn make_multi<
+    Pk: IntoDescriptorKey<Ctx>,
+    Ctx: ScriptContext,
+    V: Fn(usize, Vec<DescriptorPublicKey>) -> Terminal<DescriptorPublicKey, Ctx>,
```
Conceptual question.

The V here is a function but, in the usages of `make_multi` in dsl macros its used as, 
`$crate::keys::make_multi($thresh, $crate::miniscript::Terminal::Multi, $keys, &secp)`.

So `V` here is a `miniscript::Terminal::Multi` which is an enum..

How does this work??

afilini replied:
`Terminal::Multi` is not actually an enum, it's an enum *variant*. Just `Terminal` is the enum. Rust allows you to treat an enum variant as a function that creates an instance of that variant.

You may have used this pattern before when mapping errors since it's pretty common:

```rust
returns_a_result().map_err(MyErr::OtherVariant)
```

This is equivalent to

```rust
returns_a_result().map_err(|e| MyErr::OtherVariant(e))
```

rajarshimaitra replied:
Oh this was interesting.. Cool stuff.. Yeah I have seen that pattern before, didn't realize that's what happening here..  Thanks..

rajarshimaitra reviewed (commented):
tACK 4c50a9e37f12685b8f786cb56a25a98323a0d7a5

afilini committed:
taproot-tests: validate `tap_tree` in psbt outputs

Co-authored-by: Daniela Brozzoni <danielabrozzoni@protonmail.com>

afilini commented:
I should have addressed all the comments, here's the range-diff: https://gist.github.com/afilini/6e933f087971c83eb3b422a2a8b5f7aa

afilini commented:
I still have a few things I'd like to add to this, but I feel like this is growing a bit too much.

We should be done with the tests, how do you feel about merging this the way it is now, and adding more improvements in a follow-up PR? The main one I have in mind is a way to ask the signer to only sign for some specific leaf hashes, or for the internal key, since we currently sign for everything we can.

notmandatory commented on CHANGELOG.md:
```diff
@@ -10,8 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add traits to reuse `Blockchain`s across multiple wallets (`BlockchainFactory` and `StatelessBlockchain`).
-- Upgrade to rust-bitcoin `0.28`
 - If using the `sqlite-db` feature all cached wallet data is deleted due to a possible UTXO inconsistency, a wallet.sync will recreate it  
+- Update `PkOrF` in the policy module to become an enum
```
Should also add a note here about adding *experimental* `tr()` descriptor support, or something along those lines. :-)

rajarshimaitra replied:
+1 

Yes experimental note for having taproot capabilities would be useful.. Also its not just the `tr()` descriptor, but the full suit of taproot capability we have now in bdk right??

notmandatory replied:
:+1: These notes look good.

notmandatory reviewed (approved):
ACK 3d1f72e7459524eca07c9546a6a959401f71305f

I was only able to grok parts of this, but I agree with merging it now as *experimental* `tr()` support so we can get it into users/testers hands. This is an amazing amount of work and I hope over time I and other will be able to help @afilini more on contributing to deep changes like this. From what I've seen even bitcoin core descriptor wallets don't yet have `tr()` support that is as extensive as this PR provides.

danielabrozzoni reviewed (approved):
Core review ACK 3d1f72e7459524eca07c9546a6a959401f71305f - I reviewed the code to the best of my knowledge, and it looks correct. I haven't done any manual testing, other than running `cargo test`, but I can do that after this is merged. 

I'm hyped about taproot being merged in BDK. Nice job!! 🚀

afilini commented:
I've opened #616 so that we don't forget the extra stuff that we can add after this pr.

I'll wait for a comment from @rajarshimaitra before merging, I'm assuming he must be close to the end since he started looking at the macro stuff.

rajarshimaitra reviewed (commented):
tACK 8ce8da187ec79f04a348011bc9c76f29c0352a48

rajarshimaitra commented on CHANGELOG.md:
```diff
@@ -10,8 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - New MSRV set to `1.56`
 - Unpinned tokio to `1`
 - Add traits to reuse `Blockchain`s across multiple wallets (`BlockchainFactory` and `StatelessBlockchain`).
-- Upgrade to rust-bitcoin `0.28`
```
Why we are removing this line?

afilini replied:
Good catch, this was probably a mistake in the rebase

rajarshimaitra reviewed (approved):
tACK 3d1f72e7459524eca07c9546a6a959401f71305f

All tests and changes look good so far.. I think its good to get merged as experimental taproot capabilities..

afilini committed:
Update to the newest release of rust-bitcoin

afilini committed:
Update CHANGELOG.md for Taproot

afilini commented:
I noticed a new release of rust-bitcoin came out, I'm not sure what's new there but just in case I've updated the `Cargo.toml` to use that new version.

I've also updated the changelog with a brief list of what we support related to taproot.

notmandatory commented:
> I noticed a new release of rust-bitcoin came out, I'm not sure what's new there but just in case I've updated the `Cargo.toml` to use that new version.

Looks like the new `rust-bitcoin` 0.28.1 is only to make an exception and raise the MSRV to `1.47` for CI testing `no_std`, so probably doesn't affect `bdk` but won't hurt to specify as our min version.
